src/js/form.js

```diff
@@ -347,46 +347,96 @@ Form.prototype.init = function () {
+            const addRepeatType = events.AddRepeat().type;
+            const removeRepeatType = events.RemoveRepeat().type;
 
-        // Recalculate after repeats are initialized. Previously this was performed in
-        // `tempHandlerAddRepeat`, but recalculating them all at once is significantly faster.
-        if (repeatsAdded > 0) {
-            this.calc.update({
-                allRepeats: true,
-                cloned: true,
-            });
+            // after radio button data-name setting (now done in XLST)
+            // Set temporary event handler to ensure calculations in newly added repeats are run for the first time
+            const tempHandlerAddRepeat = ({ detail }) => {
+                const recalculate = recalculationPaths.includes(
+                    detail.repeatPath
+                );
 
-            this.all = {};
+                if (recalculate) {
+                    this.calc.update(detail);
+                    didRecalculate = true;
+                }
+            };
+            const tempHandlerRemoveRepeat = (event) => {
+                const recalculate =
+                    didRecalculate ||
+                    recalculationPaths.includes(
+                        event.detail.initRepeatInfo.repeatPath
+                    );
 
-            // Builds a cache of known repeat path prefixes `repeat.init`.
-            // The cache is sorted by length, longest to shortest, to ensure
-            // that lookups using this cache find the deepest nested repeat
-            // for a given path.
-            this.repeatPathPrefixes = Array.from(
-                this.view.html.querySelectorAll('.or-repeat-info')
-            )
-                .map((element) => `${element.dataset.name}/`)
-                .sort((a, b) => b.length - a.length);
-        }
+                if (recalculate) {
+                    this.all = {};
+                    didRecalculate = false;
+                }
+            };
 
-        this.view.html.removeEventListener(
-            events.AddRepeat().type,
-            tempHandlerAddRepeat
-        );
+            if (recalculationPaths.length > 0) {
+                this.view.html.addEventListener(
+                    addRepeatType,
+                    tempHandlerAddRepeat
+                );
+                this.view.html.addEventListener(
+                    removeRepeatType,
+                    tempHandlerRemoveRepeat
+                );
+            }
+
+            this.repeatsInitialized = true;
+            this.repeats.init();
+
+            if (recalculationPaths.length > 0) {
+                this.view.html.removeEventListener(
+                    addRepeatType,
+                    tempHandlerAddRepeat
+                );
+                this.view.html.removeEventListener(
+                    removeRepeatType,
+                    tempHandlerRemoveRepeat
+                );
+            }
+
+            if (!didRecalculate) {
```

Comment:
Remove. Might lead to extra work being done but there's some risk in having it there.